gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/InstantiatingGrpcChannelProvider.java

```diff
@@ -82,13 +81,25 @@
  * <p>The client lib header and generator header values are used to form a value that goes into the
  * http header of requests to the service.
  */
-@InternalExtensionOnly
 public final class InstantiatingGrpcChannelProvider implements TransportChannelProvider {
+
+  static String systemProductName;
+
+  static {
+    try {
+      systemProductName =
+          Files.asCharSource(new File("/sys/class/dmi/id/product_name"), StandardCharsets.UTF_8)
+              .readFirstLine();
+    } catch (IOException e) {
+      // Keep existing behavior the same (null means it is not on compute engine)
+      systemProductName = null;
+    }
+  }
```

Comment:
SystemProductName logic moved to a static block and initialized once. Stored in a variable so that it can be overriden.